libraries/chain/controller.cpp

```diff
@@ -3011,35 +3011,19 @@ block_state_ptr controller::head_block_state()const {
 }
 
 uint32_t controller::fork_db_head_block_num()const {
-   return my->fork_db.head()->block_num;
+   if( my->read_mode == db_read_mode::IRREVERSIBLE ) {
+      return my->fork_db.pending_head()->block_num;
+   } else {
+      return my->fork_db.head()->block_num;
+   }
 }
 
 block_id_type controller::fork_db_head_block_id()const {
-   return my->fork_db.head()->id;
-}
-
-time_point controller::fork_db_head_block_time()const {
-   return my->fork_db.head()->header.timestamp;
-}
-
-account_name  controller::fork_db_head_block_producer()const {
-   return my->fork_db.head()->header.producer;
-}
-
-uint32_t controller::fork_db_pending_head_block_num()const {
-   return my->fork_db.pending_head()->block_num;
-}
-
-block_id_type controller::fork_db_pending_head_block_id()const {
-   return my->fork_db.pending_head()->id;
-}
-
-time_point controller::fork_db_pending_head_block_time()const {
-   return my->fork_db.pending_head()->header.timestamp;
-}
-
-account_name  controller::fork_db_pending_head_block_producer()const {
-   return my->fork_db.pending_head()->header.producer;
+   if( my->read_mode == db_read_mode::IRREVERSIBLE ) {
+      return my->fork_db.pending_head()->id;
+   } else {
+      return my->fork_db.head()->id;
+   }
```

Comment:
It looks like we could have an internal function:

`block_state_ptr get_fork_head() { return read_mode == db_read_mode::IRREVERSIBLE) ? fork_db.pending_head() : fork_db.head(); }`

which could be used to simplify the above functions, as well as `fetch_block_state_by_number` and other locations in controller.coo.

Reply:
it sounds counter-intuitive that, when asking for IRREVERSIBLE, we return the `pending` head?

Reply:
When running in IRREVERSIBLE, "head" is irreversible; but you still are receiving blocks and tracking head of the chain.

Reply:
I thought about adding a `get_fork_head`, I didn't because I thought it was clearer to have it obvious in `fork_db_head_block_num` and `fork_db_head_block_id`. I can add it if you think it is cleaner.

Reply:
Hum. What about calling the function `get_irreversible_fork_head` to indicate it takes into account the mode?

Reply:
It is only irreversible if you are running in IRREVERSIBLE mode.

Reply:
OK, my understanding :-). `get_fork_db_head` should be in a header.
```
block_state_ptr controller::get_fork_db_head() {
   // this function returns the last block, even when read_mode is IRREVERSIBLE
   // note: In normal mode, we return head(), which is the last validated block
   //       in IRREVERSIBLE mode, we return pending_head() which is the newest block we
   //       have, and which may or may not be validated.
   return read_mode == db_read_mode::IRREVERSIBLE ?
      fork_db.pending_head() :  // should we make sure we return only validated blocks?
      fork_db.head();
}

uint32_t controller::fork_db_head_block_num()const {
   return get_fork_db_head()->block_num;
}

block_id_type controller::fork_db_head_block_id()const {
   return get_fork_db_head()->id;
}
```

Reply:
I'm not quite understanding what `pending_head` is doing? it looks like it is either returning `my-head`, just like `head()`, or a block from indx if that block has a higher `dpos_irreversible_blocknum`?

```
   block_state_ptr fork_database::pending_head()const {
      std::shared_lock g( my->mtx );
      const auto& indx = my->index.get<by_lib_block_num>();

      auto itr = indx.lower_bound( false );
      if( itr != indx.end() && !(*itr)->is_valid() ) {
         if( first_preferred( **itr, *my->head ) )
            return *itr;
      }

      return my->head;
   }
```

Reply:
The difference is `pending_head()` is returning the best/tip/preferred block that is not validated or head if head is best/tip/preferred.

Reply:
Added `fork_db_head()`